src/wmtk/operations/composite/TetEdgeSwap.cpp

```diff
@@ -0,0 +1,150 @@
+    std::vector<Tuple> face_tuples_generated_by_swap;
+
+
+    std::array<std::vector<Simplex>, 4> simplices_deleted_by_collapse;
+
+    for (const Simplex& s : sc.simplex_vector()) {
+        simplices_deleted_by_collapse[get_primitive_type_id(s.primitive_type())].emplace_back(s);
+    }
+
+    // check if #edge_generated_by_split > #edges_deleted_by_collapse, if not, return face instead
+    assert(edges_generated_by_split.size() >= simplices_deleted_by_collapse[1].size());
+
+    bool able_to_return_edges = true;
+    if (edges_generated_by_split.size() == simplices_deleted_by_collapse[1].size())
+        able_to_return_edges = false;
+
+    if (able_to_return_edges) {
+        // can use std::difference with sorted vector, worth?
+        for (const auto& e_split : edges_generated_by_split) {
+            bool preserve_flag = true;
+            for (const auto& e_collapse : simplices_deleted_by_collapse[1]) {
+                if (simplex::utils::SimplexComparisons::equal(
+                        mesh(),
+                        simplex::Simplex::edge(e_split),
+                        e_collapse)) {
+                    preserve_flag = false;
+                    break;
+                }
+            }
+            if (preserve_flag) {
+                edge_tuples_generated_by_swap.push_back(e_split);
+            }
+        }
+
+        assert(
+            edge_tuples_generated_by_swap.size() ==
+            (edges_generated_by_split.size() - simplices_deleted_by_collapse[1].size()));
+    } else {
+        // return face
+        auto faces_generated_by_split = m_split.new_face_tuples();
+        assert(faces_generated_by_split.size() >= simplices_deleted_by_collapse[2].size());
+
+        for (const auto& f_split : faces_generated_by_split) {
+            bool preserve_flag = true;
+            for (const auto& f_collapse : simplices_deleted_by_collapse[2]) {
+                if (simplex::utils::SimplexComparisons::equal(
+                        mesh(),
+                        simplex::Simplex::face(f_split),
+                        f_collapse)) {
+                    preserve_flag = false;
+                    break;
+                }
+            }
+            if (preserve_flag) {
+                face_tuples_generated_by_swap.push_back(f_split);
+            }
+        }
+
+        assert(
+            face_tuples_generated_by_swap.size() ==
+            (faces_generated_by_split.size() - simplices_deleted_by_collapse[2].size()));
+    }
+
+
+    // do collapse
+    const auto collapse_simplicies =
+        m_collapse(Simplex(m_collapse.primitive_type(), collapse_tuple));
+    if (collapse_simplicies.empty()) return {};
+    assert(collapse_simplicies.size() == 1);
+
+
+    // resurrect the edge tuples
+    if (able_to_return_edges) {
+        for (long i = 0; i < edge_tuples_generated_by_swap.size(); ++i) {
+            edge_tuples_generated_by_swap[i] = resurrect_tuple(edge_tuples_generated_by_swap[i]);
```

Comment:
did every edge tuple from the swap surrive? you need to delete one from this set right?

Reply:
why did you do two loops here?
there's also afunction for this in wmtk::simplex::utils::homgogeneous*

Reply:
Every edge in this set survives.

Got dumb for the two loops  :/